matrix_sdk/src/http_client.rs

```diff
@@ -227,46 +229,73 @@ async fn response_to_http_response(
         .expect("Can't construct a response using the given body"))
 }
 
-#[cfg(any(test, target_arch = "wasm32"))]
+#[cfg(any(target_arch = "wasm32"))]
 async fn send_request(
     client: &Client,
     request: http::Request<Vec<u8>>,
-    _: Option<Duration>,
+    _: RequestConfig,
 ) -> Result<http::Response<Vec<u8>>, HttpError> {
     let request = reqwest::Request::try_from(request)?;
     let response = client.execute(request).await?;
 
     Ok(response_to_http_response(response).await?)
 }
 
-#[cfg(all(not(test), not(target_arch = "wasm32")))]
+#[cfg(all(not(target_arch = "wasm32")))]
 async fn send_request(
     client: &Client,
     request: http::Request<Vec<u8>>,
-    timeout: Option<Duration>,
+    config: RequestConfig,
 ) -> Result<http::Response<Vec<u8>>, HttpError> {
-    let backoff = ExponentialBackoff::default();
+    let mut backoff = ExponentialBackoff::default();
     let mut request = reqwest::Request::try_from(request)?;
+    let retry_limit = Arc::new(Mutex::new(config.retry_limit));
 
-    if let Some(timeout) = timeout {
+    if let Some(timeout) = config.timeout {
         *request.timeout_mut() = Some(timeout);
     }
 
+    backoff.max_elapsed_time = config.retry_timeout;
+
     let request = &request;
+    let retry_limit = &retry_limit;
 
     let request = || async move {
+        let stop = {
+            let mut lock = retry_limit.lock().unwrap();
+            if let Some(limit) = lock.take() {
```

Comment:
Why the `take()` here instead of a mutable borrow? Changing this to a atomic u64 would remove the mutable borrow as well.

Reply:
Yeah but you don't want to modify the `Option`, you want to modify the thing that is inside the option if there is something.

```rust
if let Some(option) = option {
    *option -= 1;
}
```

Or rather use [saturating_sub()](https://doc.rust-lang.org/std/primitive.u32.html#method.saturating_sub) to avoid overflow issues.

Reply:
I removed the `take()`. Now the counter increments till it hits the limit if a limit was set.